cpp/src/io/json/json_tree.cu

```diff
@@ -254,12 +278,495 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
+                                              node_type.data() + level_boundaries[level - 1]);
+    auto adjacent_pair_it = thrust::make_zip_iterator(start_it - 1, start_it);
+    // Compares two adjacent items, beginning with the first and second item from the current level.
+    // Writes flags to the index of the rhs item.
+    // First index holds next col_id from previous level.
+    thrust::transform(rmm::exec_policy(stream),
+                      adjacent_pair_it + 1,
+                      adjacent_pair_it + level_boundaries[level] - level_boundaries[level - 1],
+                      col_id.data() + level_boundaries[level - 1] + 1,
+                      [] __device__(auto adjacent_pair) -> size_type {
+                        auto lhs = thrust::get<0>(adjacent_pair),
+                             rhs = thrust::get<1>(adjacent_pair);
+                        return lhs != rhs ? 1 : 0;
+                      });
+
+    // includes previous level last col_id to continue the index.
+    thrust::inclusive_scan(rmm::exec_policy(stream),
+                           col_id.data() + level_boundaries[level - 1],
+                           col_id.data() + level_boundaries[level] + (level != num_levels - 1),
+                           // +1 only for not-last-levels, for next level start col_id
+                           col_id.data() + level_boundaries[level - 1]);
+    PRINT_LEVEL_DATA(level, ".after scan");
+    // scatter to restore original order.
+    {
+      auto const num_nodes_per_level = level_boundaries[level] - level_boundaries[level - 1];
+      rmm::device_uvector<NodeIndexT> tmp_col_id(num_nodes_per_level, stream);
+      rmm::device_uvector<NodeIndexT> tmp_parent_col_id(num_nodes_per_level, stream);
+      thrust::scatter(rmm::exec_policy(stream),
+                      thrust::make_zip_iterator(col_id.begin() + level_boundaries[level - 1],
+                                                parent_col_id.data() + level_boundaries[level - 1]),
+                      thrust::make_zip_iterator(col_id.begin() + level_boundaries[level],
+                                                parent_col_id.data() + level_boundaries[level]),
+                      scatter_indices.begin(),
+                      thrust::make_zip_iterator(tmp_col_id.begin(), tmp_parent_col_id.begin()));
+      thrust::copy(rmm::exec_policy(stream),
+                   tmp_col_id.begin(),
+                   tmp_col_id.end(),
+                   col_id.begin() + level_boundaries[level - 1]);
+      thrust::copy(rmm::exec_policy(stream),
+                   tmp_parent_col_id.begin(),
+                   tmp_parent_col_id.end(),
+                   parent_col_id.begin() + level_boundaries[level - 1]);
+      thrust::sequence(rmm::exec_policy(stream),
+                       scatter_indices.begin(),
+                       scatter_indices.begin() + num_nodes_per_level);
+    }
```

Comment:
```suggestion
    }
    thrust::sequence(rmm::exec_policy(stream),
                     scatter_indices.begin(),
                     scatter_indices.begin() + num_nodes_per_level);
```
Just a suggestion. Move this up into the for-loop scope. Makes it a bit clearer.